components/style/properties/longhand/inherited_box.mako.rs

```diff
@@ -57,191 +57,11 @@ ${helpers.single_keyword("image-rendering",
-            } else {
-                if self.flipped {
-                    computed_value::T::AngleWithFlipped(Orientation::Angle0, true)
-                } else {
-                    computed_value::T::FromImage
-                }
-            }
-        }
-
-        #[inline]
-        fn from_computed_value(computed: &computed_value::T) -> Self {
-            match *computed {
-                computed_value::T::FromImage => SpecifiedValue { angle: None, flipped: false },
-                computed_value::T::AngleWithFlipped(ref orientation, flipped) => {
-                    SpecifiedValue {
-                        angle: Some(orientation.angle()),
-                        flipped: flipped,
-                    }
-                }
-            }
-        }
-    }
-
-    impl ToCss for computed_value::T {
-        fn to_css<W>(&self, dest: &mut W) -> fmt::Result where W: fmt::Write {
-            match *self {
-                computed_value::T::FromImage => dest.write_str("from-image"),
-                computed_value::T::AngleWithFlipped(angle, flipped) => {
-                    angle.to_css(dest)?;
-                    if flipped {
-                        dest.write_str(" flip")?;
-                    }
-                    Ok(())
-                },
-            }
-        }
-    }
-
-    // from-image | <angle> | [<angle>? flip]
-    pub fn parse<'i, 't>(context: &ParserContext, input: &mut Parser<'i, 't>)
-                         -> Result<SpecifiedValue, ParseError<'i>> {
-        if input.try(|input| input.expect_ident_matching("from-image")).is_ok() {
-            // Handle from-image
-            Ok(SpecifiedValue { angle: None, flipped: false })
-        } else if input.try(|input| input.expect_ident_matching("flip")).is_ok() {
-            // Handle flip
-            Ok(SpecifiedValue { angle: Some(Angle::zero()), flipped: true })
-        } else {
-            // Handle <angle> | <angle> flip
-            let angle = input.try(|input| Angle::parse(context, input)).ok();
-            if angle.is_none() {
-                return Err(input.new_custom_error(StyleParseErrorKind::UnspecifiedError));
-            }
-
-            let flipped = input.try(|input| input.expect_ident_matching("flip")).is_ok();
-            Ok(SpecifiedValue { angle: angle, flipped: flipped })
-        }
-    }
-</%helpers:longhand>
+${helpers.predefined_type("image-orientation",
+                          "ImageOrientation",
+                          "computed::ImageOrientation::zero()",
+                          products="gecko",
+                          animation_value_type="discrete",
+                          gecko_pref="layout.css.image-orientation.enabled",
+                          spec="https://drafts.csswg.org/css-images/#propdef-image-orientation, \
+                    /// additional values in https://developer.mozilla.org/en-US/docs/Web/CSS/image-orientation")}
```

Comment:
I had to indent this weirdly to prevent a test-tidy error due to the line length. Alternative suggestions welcome.

Reply:
How about putting the second line as a documentation for type `ImageOrientation`?

Reply:
Yeah, this is kinda weird, but it's not worse than what there was before, so feel free to leave it.

Reply:
I'm actually not sure I understand why we need the second link at all? Could we just remove it?

Reply:
Sure, we can.